File: android/src/main/java/com/supermap/rnsupermap/JSBufferAnalyst.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.supermap.analyst.BufferAnalyst;
import com.supermap.analyst.BufferAnalystGeometry;
import com.supermap.analyst.BufferAnalystParameter;
import com.supermap.analyst.BufferEndType;
import com.supermap.analyst.BufferRadiusUnit;
import com.supermap.data.Color;
import com.supermap.data.CursorType;
import com.supermap.data.Dataset;
import com.supermap.data.DatasetVector;
import com.supermap.data.Enum;
import com.supermap.data.GeoRegion;
import com.supermap.data.GeoStyle;
import com.supermap.data.Geometry;
import com.supermap.data.PrjCoordSys;
import com.supermap.data.Recordset;
import com.supermap.data.Size2D;
import com.supermap.mapping.Layer;
import com.supermap.mapping.Selection;
import com.supermap.mapping.TrackingLayer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Myself on 2017/5/22.
 */

public class JSBufferAnalyst extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSBufferAnalyst";
    protected static Map<String, BufferAnalyst> m_BufferAnalystList = new HashMap<String, BufferAnalyst>();
    BufferAnalyst m_BufferAnalyst;

    public JSBufferAnalyst(ReactApplicationContext context) {
        super(context);
    }

    public static BufferAnalyst getObjFromList(String id) {
        return m_BufferAnalystList.get(id);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @ReactMethod
    public void createBuffer(String sourceDatasetId, String resultDatasetId,
                             String bufferAnalystParamId, Boolean isUnion, Boolean isAttributeRetained, Promise promise){
        try{
            DatasetVector source = JSDatasetVector.getObjFromList(sourceDatasetId);
            DatasetVector result = JSDatasetVector.getObjFromList(resultDatasetId);
            BufferAnalystParameter bufferAnalystPara = JSBufferAnalystParameter.getObjFromList(bufferAnalystParamId);

            Boolean isCreate = BufferAnalyst.createBuffer(source,result,bufferAnalystPara,isUnion,isAttributeRetained);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isCreate",isCreate);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void createLineOneSideMultiBuffer(String sourceDatasetId, String resultDatasetId,
                                             ReadableArray arrBufferRadius, int bufferRadiusUnit, int semicircleSegment, Boolean isLeft, Boolean isUnion, Boolean isAttributeRetained, Boolean isRing, Promise promise){
        try{
            DatasetVector source = JSDatasetVector.getObjFromList(sourceDatasetId);
            DatasetVector result = JSDatasetVector.getObjFromList(resultDatasetId);
            BufferRadiusUnit unit = (BufferRadiusUnit) Enum.parse(BufferRadiusUnit.class,bufferRadiusUnit);

            ArrayList listArr = arrBufferRadius.toArrayList();
            Object[] objArr = listArr.toArray();
            double[] doubleArr = new double[objArr.length];
            for (int i = 0; i<=objArr.length-1;i++){
                doubleArr[i] = (double)objArr[i];
            }
            Boolean isCreate = BufferAnalyst.createLineOneSideMultiBuffer(source,result,doubleArr,unit,semicircleSegment,isLeft,isUnion,isAttributeRetained,isRing);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isCreate",isCreate);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void createMultiBuffer(String sourceDatasetId, String resultDatasetId,
                                  ReadableArray arrBufferRadius, int bufferRadiusUnit, int semicircleSegment, Boolean isUnion, Boolean isAttributeRetained,Boolean isRing, Promise promise){
        try{
            DatasetVector source = JSDatasetVector.getObjFromList(sourceDatasetId);
            DatasetVector result = JSDatasetVector.getObjFromList(resultDatasetId);
            BufferRadiusUnit unit = (BufferRadiusUnit) Enum.parse(BufferRadiusUnit.class,bufferRadiusUnit);

            ArrayList listArr = arrBufferRadius.toArrayList();
            Object[] objArr = listArr.toArray();
            double[] doubleArr = new double[objArr.length];
            for (int i = 0; i<=objArr.length-1;i++){
                doubleArr[i] = (double)objArr[i];
            }
            Boolean isCreate = BufferAnalyst.createMultiBuffer(source,result,doubleArr,unit,semicircleSegment,isUnion,isAttributeRetained,isRing);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isCreate",isCreate);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void analyst(String mapId, String layerId, ReadableMap params, Promise promise){
        try{
            com.supermap.mapping.Map map = JSMap.getObjFromList(mapId);
            Layer layer = JSLayer.getLayer(layerId);
            Selection selection = layer.getSelection();
            Recordset recordset = selection.toRecordset();

            TrackingLayer trackingLayer = map.getTrackingLayer();

            trackingLayer.clear();
            while (!recordset.isEOF()) {

                Geometry geoForBuffer = recordset.getGeometry();
                Dataset queryDataset = recordset.getDataset();

                PrjCoordSys prjCoordSys = queryDataset.getPrjCoordSys();

                GeoStyle geoStyle = new GeoStyle();

                BufferAnalystParameter bufferAnalystParameter = new BufferAnalystParameter();
                ReadableMap parameter = params.getMap("parameter");

                if (params.hasKey("parameter")) {
                    if (parameter.hasKey("endType")) {
                        BufferEndType endType = (BufferEndType) Enum.parse(BufferEndType.class, parameter.getInt("endType"));
                        bufferAnalystParameter.setEndType(endType);
                    }
                    if (parameter.hasKey("leftDistance")) {
                        bufferAnalystParameter.setLeftDistance(parameter.getInt("leftDistance"));
                    }
                    if (parameter.hasKey("rightDistance")) {
                        bufferAnalystParameter.setRightDistance(parameter.getInt("rightDistance"));
                    }
                }

                if (params.hasKey("geoStyle")) {
                    ReadableMap geoStyleMap = params.getMap("geoStyle");

                    if (geoStyleMap.hasKey("lineWidth")) {
                        geoStyle.setLineWidth(geoStyleMap.getInt("lineWidth"));
                    }
                    if (geoStyleMap.hasKey("fillForeColor")) {
                        ReadableMap fillForeColor = geoStyleMap.getMap("fillForeColor");
                        geoStyle.setFillForeColor(new Color(fillForeColor.getInt("r"), fillForeColor.getInt("g"), fillForeColor.getInt("b")));
                    }
                    if (geoStyleMap.hasKey("lineColor")) {
                        ReadableMap lineColor = geoStyleMap.getMap("lineColor");
                        geoStyle.setLineColor(new Color(lineColor.getInt("r"), lineColor.getInt("g"), lineColor.getInt("b")));
                    }
                    if (geoStyleMap.hasKey("lineSymbolID")) {
                        geoStyle.setLineSymbolID(geoStyleMap.getInt("lineSymbolID"));
                    }
                    if (geoStyleMap.hasKey("markerSymbolID")) {
                        geoStyle.setMarkerSymbolID(geoStyleMap.getInt("markerSymbolID"));
                    }
                    if (geoStyleMap.hasKey("markerSize")) {
                        ReadableMap markerSize = geoStyleMap.getMap("markerSize");
                        geoStyle.setMarkerSize(new Size2D(markerSize.getInt("w"), markerSize.getInt("h")));
                    }
                    if (geoStyleMap.hasKey("fillOpaqueRate")) {
                        geoStyle.setMarkerSymbolID(geoStyleMap.getInt("fillOpaqueRate"));
                    }
                }

                GeoRegion geoRegion = BufferAnalystGeometry.createBuffer(geoForBuffer, bufferAnalystParameter, prjCoordSys);
                geoRegion.setStyle(geoStyle);

                trackingLayer.add(geoRegion, "");

                recordset.moveNext();
            }
            recordset.dispose();

            map.refresh();
            promise.resolve(true);
        }catch (Exception e){
            e.printStackTrace();
            promise.reject(e);
        }
    }
}

File: ios/oldNative/JSBufferAnalyst.m
//
//  JSBufferAnalyst.m
//  Supermap
//
//  Created by 王子豪 on 2017/5/19.
//  Copyright © 2017年 Facebook. All rights reserved.
//
#import "JSBufferAnalyst.h"
#import "SuperMap/BufferAnalyst.h"
#import "SuperMap/Map.h"
#import "SuperMap/Selection.h"
#import "SuperMap/TrackingLayer.h"
#import "SuperMap/Recordset.h"
#import "SuperMap/PrjCoordSys.h"
#import "SuperMap/DatasetVector.h"
#import "SuperMap/BufferAnalystParameter.h"
#import "SuperMap/BufferAnalystGeometry.h"
#import "SuperMap/GeoStyle.h"
#import "SuperMap/Color.h"
#import "SuperMap/Size2D.h"
#import "SuperMap/GeoRegion.h"
#import "SuperMap/Layer.h"


#import "JSObjManager.h"
@implementation JSBufferAnalyst
RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(createBuffer,createBufferBySourceId:(NSString*)sourceId resultId:(NSString*)resultId bufferAnalystParaId:(NSString*)paraId isUnion:(BOOL)isUnion isAttributeRetained:(BOOL)isAttributeRetained resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        DatasetVector* sourceDataSet = [JSObjManager getObjWithKey:sourceId];
        DatasetVector* resultDataSet = [JSObjManager getObjWithKey:resultId];
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        BOOL create = [BufferAnalyst createBufferSourceVector:sourceDataSet ResultVector:resultDataSet BufferParam:para IsUnion:isUnion IsAttributeRetained:isAttributeRetained];
        NSNumber* nsCreate = [NSNumber numberWithBool:create];
        resolve(@{@"isCreate":nsCreate});
    } @catch (NSException *exception) {
        reject(@"BufferAnalyst",@"create Buffer exception",nil);
    }
}

RCT_REMAP_METHOD(createLineOneSideMultiBuffer,createLineOneSideMultiBufferBySourceId:(NSString*)sourceId resultId:(NSString*)resultId arrBufferRadius:(NSArray*)arrBufferRadius bufferRadiusUnit:(int)bufferRadiusUnit semicircleSegment:(int)semicircleSegment isLeft:(BOOL)isLeft isUnion:(BOOL)isUnion isAttributeRetained:(BOOL)isAttributeRetained isRing:(BOOL)isRing resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        DatasetVector* sourceDataSet = [JSObjManager getObjWithKey:sourceId];
        DatasetVector* resultDataSet = [JSObjManager getObjWithKey:resultId];
        BOOL create = [BufferAnalyst createLineOneSideMultiBufferSourceVector:sourceDataSet ResultVector:resultDataSet ArrBufferRadius:arrBufferRadius BufferRadiusUnit:bufferRadiusUnit SemicircleSegment:semicircleSegment IsLeft:isLeft IsUnion:isUnion IsAttributeRetained:isAttributeRetained IsRing:isRing];
        NSNumber* nsCreate = [NSNumber numberWithBool:create];
        resolve(@{@"isCreate":nsCreate});
    } @catch (NSException *exception) {
        reject(@"BufferAnalyst",@"create Line One Side MultiBuffer exception",nil);
    }
}

RCT_REMAP_METHOD(createMultiBuffer,createMultiBufferBySourceId:(NSString*)sourceId resultId:(NSString*)resultId arrBufferRadius:(NSArray*)arrBufferRadius bufferRadiusUnit:(int)bufferRadiusUnit semicircleSegment:(int)semicircleSegment isUnion:(BOOL)isUnion isAttributeRetained:(BOOL)isAttributeRetained isRing:(BOOL)isRing resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        DatasetVector* sourceDataSet = [JSObjManager getObjWithKey:sourceId];
        DatasetVector* resultDataSet = [JSObjManager getObjWithKey:resultId];
        BOOL create = [BufferAnalyst createMultiBufferSourceVector:sourceDataSet ResultVector:resultDataSet ArrBufferRadius:arrBufferRadius BufferRadiusUnit:bufferRadiusUnit SemicircleSegment:semicircleSegment IsUnion:isUnion IsAttributeRetained:isAttributeRetained IsRing:isRing];
        NSNumber* nsCreate = [NSNumber numberWithBool:create];
        resolve(@{@"isCreate":nsCreate});
    } @catch (NSException *exception) {
        reject(@"BufferAnalyst",@"create Multi Buffer exception",nil);
    }
}

RCT_REMAP_METHOD(analyst, analystByMapId:(NSString*)mapId layerId:(NSString*)layerId params:(NSDictionary*)params  resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        Map* map = [JSObjManager getObjWithKey:mapId];
        Layer* layer = [JSObjManager getObjWithKey:layerId];
        Selection* selection = [layer getSelection];
        Recordset* recordset = [selection toRecordset];
        TrackingLayer* trackingLayer = map.trackingLayer;
        
        [trackingLayer clear];
        
        while (!recordset.isEOF) {
            Geometry* geoForBuffer = recordset.geometry;
            DatasetVector* queryDataset = recordset.datasetVector;
            
            PrjCoordSys* prj = queryDataset.prjCoordSys;
            
            GeoStyle* geoStyle = [[GeoStyle alloc] init];
            BufferAnalystParameter* bufferAnalystParameter = [[BufferAnalystParameter alloc] init];
            NSDictionary* parameter = [params objectForKey:@"parameter"];
            if (parameter) {
                if ([parameter objectForKey:@"endType"]) {
                    bufferAnalystParameter.bufferEndType = (int)[parameter objectForKey:@"endType"];
                }
                if ([parameter objectForKey:@"leftDistance"]) {
                    NSNumber* leftDistance = [parameter objectForKey:@"leftDistance"];
                    bufferAnalystParameter.leftDistance = [leftDistance stringValue];
                }
                if ([parameter objectForKey:@"rightDistance"]) {
                    NSNumber* rightDistance = [parameter objectForKey:@"rightDistance"];
                    bufferAnalystParameter.rightDistance = [rightDistance stringValue];
                }
            }
            NSDictionary* geoStyleDic = [params objectForKey:@"geoStyle"];
            if (geoStyleDic) {
                if ([geoStyleDic objectForKey:@"lineWidth"]) {
                    NSNumber* lineWidthObj = [geoStyleDic objectForKey:@"lineWidth"];
                    [geoStyle setLineWidth: lineWidthObj.doubleValue];
                }
                if ([geoStyleDic objectForKey:@"fillForeColor"]) {
                    NSDictionary* fillForeColor = [geoStyleDic objectForKey:@"fillForeColor"];
                    NSNumber* r = [fillForeColor objectForKey:@"r"];
                    NSNumber* g = [fillForeColor objectForKey:@"g"];
                    NSNumber* b = [fillForeColor objectForKey:@"b"];
                    NSNumber* a = [fillForeColor objectForKey:@"a"];
                    
                    if (!a) {
                        [geoStyle setFillForeColor:[[Color alloc] initWithR:r.intValue G:g.intValue B:b.intValue]];
                    } else {
                        [geoStyle setFillForeColor:[[Color alloc] initWithR:r.intValue G:g.intValue B:b.intValue A:a.intValue]];
                    }
                    
                }
                if ([geoStyleDic objectForKey:@"lineColor"]) {
                    NSDictionary* lineColor = [geoStyleDic objectForKey:@"lineColor"];
                    NSNumber* r = [lineColor objectForKey:@"r"];
                    NSNumber* g = [lineColor objectForKey:@"g"];
                    NSNumber* b = [lineColor objectForKey:@"b"];
                    NSNumber* a = [lineColor objectForKey:@"a"];
                    
                    if (!a) {
                        [geoStyle setLineColor:[[Color alloc] initWithR:r.intValue G:g.intValue B:b.intValue]];
                    } else {
                        [geoStyle setLineColor:[[Color alloc] initWithR:r.intValue G:g.intValue B:b.intValue A:a.intValue]];
                    }
                }
                if ([geoStyleDic objectForKey:@"lineSymbolID"]) {
                    NSNumber* lineSymbolID = [geoStyleDic objectForKey:@"lineSymbolID"];
                    [geoStyle setLineSymbolID:lineSymbolID.intValue];
                }
                if ([geoStyleDic objectForKey:@"markerSymbolID"]) {
                    NSNumber* markerSymbolID = [geoStyleDic objectForKey:@"markerSymbolID"];
                    [geoStyle setLineSymbolID:markerSymbolID.intValue];
                }
                if ([geoStyleDic objectForKey:@"markerSize"]) {
                    NSDictionary* markerSize = [geoStyleDic objectForKey:@"markerSize"];
                    NSNumber* w = [markerSize objectForKey:@"w"];
                    NSNumber* h = [markerSize objectForKey:@"h"];
                    
                    Size2D* size2D = [[Size2D alloc] initWithWidth:w.doubleValue Height:h.doubleValue];
                    [geoStyle setMarkerSize:size2D];
                }
                if ([geoStyleDic objectForKey:@"fillOpaqueRate"]) {
                    NSNumber* fillOpaqueRate = [geoStyleDic objectForKey:@"fillOpaqueRate"];
                    [geoStyle setFillOpaqueRate:fillOpaqueRate.intValue];
                }
            }
            GeoRegion* geoRegion = [BufferAnalystGeometry CreateBufferSourceGeometry:geoForBuffer BufferParam:bufferAnalystParameter prjCoordSys:prj];
            [geoRegion setStyle:geoStyle];
            
            [trackingLayer addGeometry:geoRegion WithTag:@""];
            [recordset moveNext];
        }
        
        [recordset dispose];
        [map refresh];
        
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
        reject(@"BufferAnalyst", exception.reason, nil);
    }
}
@end
